### PR TITLE
NEXUS-4628: everything else aside of perf tests

### PR DESCRIPTION
**Pull request by cstamas**
This pull contains important refactoring of core attribute components and other:
- refactoring of AttributeStorages removing code dupe
- centralizing "lastRequested" tweak settings and adding new ones (see wiki: https://docs.sonatype.com/display/Nexus/Configuration)
- the "perf test" components brought into test scope, but dormant

Meaning, "everything is here", but the problematic Jackson stuff is "dormant" (ignored) for now.

### CODE REVIEW

File: nexus/nexus-proxy/src/main/java/org/sonatype/nexus/proxy/attributes/DefaultAttributesHandler.java [hunk 1]
```diff
@@ -438,4 +510,25 @@ public class DefaultAttributesHandler
...
     }
 
+    // ==
+
+    /**
+     * For UT access!
+     * 
+     * @return
+     */
+    public long getLastRequestedResolution()
```
Comment (nabcos) [inline nexus/nexus-proxy/src/main/java/org/sonatype/nexus/proxy/attributes/DefaultAttributesHandler.java:520]: if it's for UTs make it package scope?

Comment (cstamas) [inline nexus/nexus-proxy/src/main/java/org/sonatype/nexus/proxy/attributes/DefaultAttributesHandler.java:520]: Well, yes and no: yes UT uses it, but also does the "perf" test, that is in separate package...

File: nexus/nexus-proxy/pom.xml [hunk 2]
```diff
@@ -185,6 +186,19 @@
       <artifactId>shiro-web</artifactId>
       <scope>test</scope>
     </dependency>
+    
+    <!-- NEXUS-4628: Attributes perf tests, for now the "alternate" imples are in tests not getting into core -->
```
Comment (peterlynch) [inline nexus/nexus-proxy/pom.xml:190]: thx for the comment

File: nexus/nexus-proxy/src/main/java/org/sonatype/nexus/proxy/attributes/DefaultAttributesHandler.java (hunk 1 above)
Comment (peterlynch) [inline nexus/nexus-proxy/src/main/java/org/sonatype/nexus/proxy/attributes/DefaultAttributesHandler.java:520]: @VisibleForTesting ?
http://docs.guava-libraries.googlecode.com/git-history/v10.0.1/javadoc/com/google/common/annotations/VisibleForTesting.html

File: nexus/nexus-proxy/src/main/java/org/sonatype/nexus/proxy/attributes/DelegatingAttributeStorage.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+ * "fallback" to some other instance. Usable for scenarios where "transitioning" (smooth upgrade for example) is to be
+ * used, the "main" attribute storage would be "upgraded" from "legacy" attribute storage as the attributes are
+ * requested over the time from this instance.
+ * 
+ * @author cstamas
+ */
+// @Typed( AttributeStorage.class )
+// @Named( "delegating" )
+// @Singleton
+public class DelegatingAttributeStorage
```
Comment (nabcos) [inline nexus/nexus-proxy/src/main/java/org/sonatype/nexus/proxy/attributes/DelegatingAttributeStorage.java:37]: Is there a value in keeping this around when it's deactivated?

File: nexus/nexus-proxy/src/test/java/org/sonatype/nexus/proxy/attributes/perf/internal/OrderedRunner.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+        Collections.sort( copy, new Comparator<FrameworkMethod>()
+        {
+            public int compare( FrameworkMethod o1, FrameworkMethod o2 )
+            {
+                return o1.getName().compareTo( o2.getName() );
+            }
+        } );
+        return copy;
+    }
+}
```
Comment (peterlynch) [inline nexus/nexus-proxy/src/test/java/org/sonatype/nexus/proxy/attributes/perf/internal/OrderedRunner.java:50]: This class may be better suited in nexus-test-common or sisu-litmus. But I am not going to -1 this.

**Comment by peterlynch** [comment]
+1 thx for adding to https://docs.sonatype.com/display/Nexus/Configuration

**Comment by nabcos** [comment]
+1

**Comment by cstamas** [comment]
Spawned CI "feature" build:
https://builds.sonatype.org/view/nexus/job/nexus-oss-its-feature/63/

But Win slave is in bad shape it seems....

**Comment by cstamas** [comment]
Rebase happened against master, since it seems this branch picked up some problems from it:
https://builds.sonatype.org/view/nexus/job/nexus-oss-its-feature/61/

And build no63 spawned now will build the same (did not notice that push was rejected since rebase happened, and did not issue `-f`) :bomb:

**Comment by cstamas** [comment]
Build job no63 killed (I just noticed win slave was `0xdeadbeef` with it already), and spawned new one against now properly rebased branch:
https://builds.sonatype.org/view/nexus/job/nexus-oss-its-feature/64/

**Comment by velo** [comment]
I think CI just voted -7 for this =/
https://builds.sonatype.org/view/nexus/job/nexus-oss-its-feature/64/testReport/

Changes seems ok so +1, just please do not merge this before CI passes

**Comment by cstamas** [comment]
WTF

1 Caused by: java.lang.ClassNotFoundException: org.sonatype.nexus.proxy.http.HttpRequest
2 NoSuchElement (plexus did not found component) role: org.sonatype.jettytestsuite.ServletServer
3 java.lang.NoClassDefFoundError: org/sonatype/nexus/proxy/attributes/perf/internal/OrderedRunner

These are the causes of those 7 failures. The problems are:

1 is in nexus codebase
2 almost all of the proxy tests use this class, why only this one test failed if not present?
3 is in nexus codebase.

So, this is really a WTF

**Comment by cstamas** [comment]
Ok, I am merging this pull req:
- CI did a +0.5 (win slave is dead): https://builds.sonatype.org/view/nexus/job/nexus-oss-its-feature/66/
- while out, I triggered a build with `-Dit=true` and it finished okay

**Pull request merged**